__tests__/client/service-worker/events/utility.spec.js

```diff
@@ -0,0 +1,241 @@
+    const meta = { ...metaBluePrint };
+    expect(markResourceForRemoval(metaBluePrint, meta)).toBe(false);
+    [
+      // test bundling triggers change
+      ['bundle', 'legacy'],
+      // version change, whether up or down, will invalidate
+      ['version', '4.5.6'],
+      // only a single locale and lang-pack per cache
+      ['locale', 'en-CA'],
+      // if the clientCacheRevision has changed, we should update
+      ['revision', '42'],
+      // we are not invalidating for cacheName changes
+      ['cacheName', 'change-cache', false],
+    ].forEach(([propName, value, result = true]) => {
+      // set the value for a given prop name to validate
+      meta[propName] = value;
+      // run validation and observe expected result
+      expect(markResourceForRemoval(metaBluePrint, meta)).toBe(result);
+      // reset the property to match the metaBluePrint
+      meta[propName] = metaBluePrint[propName];
+    });
+  });
+});
+
+describe(createResourceMetaData.name, () => {
+  const appInfo = ['app', 'https://example.com/cdn/app/1.2.3-rc.4-abc123/'];
+  const appMetaData = {
+    type: 'one-app',
+    cacheName: 'one-app',
+    name: 'app',
+    bundle: 'browser',
+    version: '1.2.3-rc.4-abc123',
+  };
+  const moduleInfo = ['module', 'https://example.com/cdn/modules/test-root/2.2.2/'];
+  const baseMetaData = {
+    bundle: 'browser',
+    type: 'modules',
+    name: 'module',
+    version: '2.2.2',
+    cacheName: 'modules',
+    revision: '101010',
+  };
+  test.each([
+    // app
+    ['https://example.com/cdn/app/1.2.3-rc.4-abc123/app.js', appInfo, {
+      ...appMetaData,
+      path: 'app.js',
+    }],
+    ['https://example.com/cdn/app/1.2.3-rc.4-abc123/i18n/en-US.js', appInfo, {
+      ...appMetaData,
+      path: 'i18n/language.js',
+      locale: 'en-US',
+    }],
+    // modules
+    ['https://example.com/cdn/modules/test-root/2.2.2/test-root.browser.js', moduleInfo, {
+      ...baseMetaData,
+      path: 'test-root.browser.js',
+    }],
+    ['https://example.com/cdn/modules/test-root/2.2.2/test-root.legacy.browser.js', moduleInfo, {
+      ...baseMetaData,
+      bundle: 'legacy',
+      path: 'test-root.legacy.browser.js',
+    }],
+    ['https://example.com/cdn/modules/test-root/2.2.2/locale/en-US/test-root.json', moduleInfo, {
+      ...baseMetaData,
+      type: 'lang-packs',
+      path: 'en-US/test-root.json',
+      cacheName: 'lang-packs',
+      locale: 'en-US',
+    }],
+    ['https://example.com/cdn/modules/test-root/2.2.2/test-root.browser.js', moduleInfo, {
+      ...baseMetaData,
+      path: 'test-root.browser.js',
+      revision: null,
+    }],
+  ])('extracts metadata from %s', (url, [type, baseUrl], result) => {
+    // eslint-disable-next-line no-param-reassign
+    if (result.revision === null) delete result.revision;
+    const meta = createResourceMetaData({ request: { url } }, [type, baseUrl], result.revision);
+    expect(meta).toEqual({ ...result, url });
```

Comment:
Sorry, this seems overly abstracted for a test, I am struggling to figure out given X the result should be Y. You can make tests as verbose as they need to be so it clearly shows the api is doing whats expected.

Reply:
Valid point, I'll update this.

Reply:
Updated (c48473f)